// ngx-video-list-player/esm2020/lib/ngx-video-list-player.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgxVideoListPlayerComponent } from './Components/ngx-video-list-player.component';
import { StopPropagationDirective } from './Directives/stop-propagation.directive';
import { YouTubePlayerModule } from '@angular/youtube-player';
import * as i0 from "@angular/core";
export class NgxVideoListPlayerModule {
}
NgxVideoListPlayerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: NgxVideoListPlayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxVideoListPlayerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: NgxVideoListPlayerModule, declarations: [NgxVideoListPlayerComponent,
        StopPropagationDirective], imports: [CommonModule,
        YouTubePlayerModule], exports: [NgxVideoListPlayerComponent] });
NgxVideoListPlayerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: NgxVideoListPlayerModule, imports: [[
            CommonModule,
            YouTubePlayerModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: NgxVideoListPlayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxVideoListPlayerComponent,
                        StopPropagationDirective
                    ],
                    imports: [
                        CommonModule,
                        YouTubePlayerModule
                    ],
                    exports: [
                        NgxVideoListPlayerComponent
                    ]
                }]
        }] });
