// @poseclop/ngx-video-player/fesm2022/poseclop-ngx-video-player.mjs
import * as i0 from '@angular/core';
import { EventEmitter, HostListener, Output, Directive, Pipe, ViewChild, Input, Component, NgModule } from '@angular/core';
import { Subject, BehaviorSubject, combineLatest, takeUntil, fromEvent, tap, debounceTime } from 'rxjs';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';

class ClickOutsideDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.clickOutside = new EventEmitter();
    }
    onClick(target) {
        const clickedInside = this.elementRef.nativeElement.contains(target);
        if (!clickedInside) {
            this.clickOutside.emit();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: ClickOutsideDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "19.2.5", type: ClickOutsideDirective, isStandalone: false, selector: "[clickOutside]", outputs: { clickOutside: "clickOutside" }, host: { listeners: { "document:click": "onClick($event.target)" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: ClickOutsideDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[clickOutside]',
                    standalone: false
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }], propDecorators: { clickOutside: [{
                type: Output
            }], onClick: [{
                type: HostListener,
                args: ['document:click', ['$event.target']]
            }] } });

class TimePipe {
    transform(value) {
        const hours = Math.floor(value / 3600);
        const minutes = Math.floor((value - hours * 3600) / 60);
        const minuteString = hours ? minutes < 10 ? `0${minutes}` : `${minutes}` : `${minutes}`;
        const seconds = Math.floor(value - hours * 3600 - minutes * 60);
        const secondString = seconds < 10 ? `0${seconds}` : `${seconds}`;
        return `${hours ? hours : ''}${hours ? ':' : ''}${minuteString}:${secondString}`;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: TimePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "19.2.5", ngImport: i0, type: TimePipe, isStandalone: false, name: "time" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: TimePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'time',
                    standalone: false
                }]
        }] });

class ChapterPipe {
    transform(seconds, chapters) {
        const chapter = [...chapters]
            .sort((a, b) => b.time - a.time)
            .find(chapter => seconds >= chapter.time);
        return chapter ? chapter.title : '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: ChapterPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "19.2.5", ngImport: i0, type: ChapterPipe, isStandalone: false, name: "chapter" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: ChapterPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'chapter',
                    standalone: false
                }]
        }] });

class SortPipe {
    transform(values) {
        return values.sort((a, b) => a.time - b.time);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: SortPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "19.2.5", ngImport: i0, type: SortPipe, isStandalone: false, name: "sort" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: SortPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sort',
                    standalone: false
                }]
        }] });

/* eslint-disable @angular-eslint/no-output-native */
/**
 * A video player component for Angular
 */
class NgxVideoPlayerComponent {
    constructor() {
        //#region INPUTS
        /** The source of the poster that will be used for the video */
        this.posterSrc = '';
        /** The chapters for the video */
        this.chapters = [{ title: '', time: 0 }];
        /** The tracks for the video */
        this.tracks = [];
        this.controls = false;
        /** Autoplay the video */
        this.autoplay = false;
        /** Expose video loop attribute */
        this.loop = false;
        /** The volume of the video */
        this.volume = 0.5;
        //#endregion
        //#region PUBLIC PROPERTIES
        /** Does the browser support video */
        this.browserSupportsVideo = !!document.createElement('video').canPlayType;
        /** Is fullScreen enabled */
        this.fullScreenEnabled = !!document.fullscreenEnabled;
        /** The video sources (Used only internaly. Use [sources] to update video sources when interacting with the component) */
        this._videoSources = [];
        /** Is the track menu open */
        this.trackMenuOpen = false;
        /** Is the mouse moving (will be set to false only after specified delay) */
        this.mouseMoving = false;
        this.isNaN = isNaN;
        //#endregion
        //#region PRIVATE PROPERTIES
        /** The unsubscribe subject emits and complete on destroy */
        this.unsubscribe$ = new Subject();
        /** The update Thumbnail subject */
        this.updateThumbnail$ = new Subject();
        /** The reset Thumbnail subject */
        this.resetThumbnail$ = new BehaviorSubject(false);
    }
    /** The sources for the video */
    set sources(sources) {
        this._videoSources = sources;
        if (this.video) {
            this.video.nativeElement.load();
        }
    }
    /** Use optionally to add a single source to the video */
    set src(src) {
        this._src = src;
        if (this.video) {
            this.video.nativeElement.src = src;
        }
    }
    /** Play/Pause the video */
    set play(value) {
        if (value) {
            this.video.nativeElement.play();
        }
        else {
            this.video.nativeElement.pause();
        }
    }
    /** Expose the video buffered attribute */
    get buffered() {
        return this.video.nativeElement.buffered;
    }
    //#endregion
    //#region LIFECYCLE HOOKS
    ngOnInit() {
        this.handleThumbnailDisplay();
    }
    ngAfterViewInit() {
        this.handleMouseMovement();
        if (this.height) {
            this.video.nativeElement.height = this.height;
        }
        if (this.width) {
            this.video.nativeElement.width = this.width;
        }
        if (this._src) {
            this.video.nativeElement.src = this._src;
            this.video.nativeElement.load();
        }
    }
    ngOnDestroy() {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    }
    //#endregion
    //#region PUBLIC METHODS
    /**
     * Toggles the video play/pause state
     */
    toggleVideoPlayPause() {
        if (this.video.nativeElement.paused || this.video.nativeElement.ended) {
            this.video.nativeElement.play();
        }
        else {
            this.video.nativeElement.pause();
        }
    }
    /**
     * Stops the video and resets the time to 0
     */
    stopVideo() {
        this.video.nativeElement.pause();
        this.video.nativeElement.currentTime = 0;
    }
    /**
     * Sets the video time to the specified number of seconds
     * @param seconds The number of seconds to set the video time to
     */
    setVideoTime(seconds) {
        this.video.nativeElement.currentTime = seconds;
    }
    /**
     * Sets the video volume to the specified number
     * @param volume The volume to set the video to
     */
    setVideoVolume(volume) {
        this.video.nativeElement.volume = volume;
    }
    /**
     * Sets the video subtitle track to use if any was selected
     * @param track The track to select
     */
    selectSubtitles(track) {
        for (let i = 0; i < this.video.nativeElement.textTracks.length; i++) {
            if (track === null) {
                this.video.nativeElement.textTracks[i].mode = 'hidden';
            }
            else if (this.video.nativeElement.textTracks[i].language === track.srclang) {
                this.video.nativeElement.textTracks[i].mode = 'showing';
            }
            else {
                this.video.nativeElement.textTracks[i].mode = 'hidden';
            }
        }
        this.trackMenuOpen = false;
    }
    /**
     * Toggle the fullScreen mode
     */
    toggleFullScreen() {
        if (document.fullscreenElement) {
            document.exitFullscreen();
        }
        else {
            this.figure.nativeElement.requestFullscreen();
        }
    }
    /**
     * Advances the video by the specified number of seconds
     * @param seconds The number of seconds to advance the video by
     */
    advanceVideoBy(seconds) {
        this.setVideoTime(this.video.nativeElement.currentTime + seconds);
    }
    //#endregion
    //#region PRIVATE METHODS
    /**
     * Handles the display of the progress thumbnail
     */
    handleThumbnailDisplay() {
        let video;
        let canvas;
        let chapters;
        combineLatest([this.updateThumbnail$, this.resetThumbnail$])
            .pipe(takeUntil(this.unsubscribe$))
            .subscribe(([seconds, resetThumbnail]) => {
            if (!video || resetThumbnail) {
                video = this.video.nativeElement.cloneNode(true);
                video.muted = true;
                video.autoplay = false;
                video.crossOrigin = 'anonymous';
                video.addEventListener('seeked', () => {
                    try {
                        canvas
                            .getContext('2d')
                            ?.clearRect(0, 0, canvas.width, canvas.height);
                        canvas
                            .getContext('2d')
                            ?.drawImage(video, 0, 0, canvas.width, canvas.height);
                        this.thumbnailSrc = canvas.toDataURL();
                    }
                    catch (error) {
                        console.error("Error drawing thumbnail. It's likely that CORS headers are not present on the video", error);
                        this.thumbnailSrc = '';
                    }
                });
            }
            if (!canvas || resetThumbnail) {
                canvas = document.createElement('canvas');
                const ratio = this.video.nativeElement.videoWidth /
                    this.video.nativeElement.videoHeight;
                canvas.width = 160;
                canvas.height = Math.floor(160 / ratio);
            }
            if (!chapters || resetThumbnail) {
                chapters = this.chapters
                    ? [...this.chapters].sort((a, b) => b.time - a.time)
                    : [];
            }
            this.thunmnailChapter = chapters.find((chapter) => chapter.time <= seconds)?.title;
            this.thumbnailTime = seconds;
            if (video && video.readyState >= 2) {
                video.currentTime = seconds;
            }
            if (resetThumbnail) {
                this.resetThumbnail$.next(false);
            }
        });
    }
    /**
     * Handles the display of the video controls on mouse movement
     */
    handleMouseMovement() {
        fromEvent(this.figure.nativeElement, 'mousemove')
            .pipe(tap(() => {
            this.mouseMoving = true;
        }), debounceTime(1500), tap(() => {
            this.mouseMoving = false;
        }), takeUntil(this.unsubscribe$))
            .subscribe();
    }
    //#endregion
    /**
     * Coerces a data-bound value (typically a string) to a boolean.
     * @param value The value to coerce
     * @returns The coerced value
     */
    _coerceBooleanProperty(value) {
        return value != null && `${value}` !== 'false';
    }
    //#region EVENT HANDLERS
    /**
     * Called when the value of volume input is changed
     * @param event The event
     */
    onVolumeChange(event) {
        const volume = event.target.value;
        this.setVideoVolume(+volume);
    }
    /**
     * Called when the metadata of the video is loaded
     */
    onMetadataLodaded() {
        // Sort chapters by time and calculate duration
        this.chapters = this.chapters
            .sort((a, b) => a.time - b.time)
            .map((chapter, index) => ({
            ...chapter,
            duration: (this.chapters[index + 1]?.time ||
                this.video.nativeElement.duration) - chapter.time,
        }));
        this.resetThumbnail$.next(true);
    }
    /**
     * Called when the user hovers over the progress bar
     * @param event The mouse event
     * @returns void
     */
    onProgressHover(event) {
        if (!event)
            return;
        const rect = this.progress.nativeElement.getBoundingClientRect();
        const pos = (event.clientX - rect.left) / rect.width;
        this.progress.nativeElement.parentElement?.style.setProperty('--hover-x', `${event.clientX - rect.left}px`);
        this.updateThumbnail$.next(Math.floor(pos * this.video.nativeElement.duration));
    }
    /**
     * Sets the video time based on the progress bar position
     * @param event The mouse event
     */
    onProgressBarClick(event) {
        const rect = this.progress.nativeElement.getBoundingClientRect();
        const pos = (event.clientX - rect.left) / rect.width;
        this.setVideoTime(pos * this.video.nativeElement.duration);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: NgxVideoPlayerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "19.2.5", type: NgxVideoPlayerComponent, isStandalone: false, selector: "ngx-video-player", inputs: { posterSrc: "posterSrc", chapters: "chapters", tracks: "tracks", sources: "sources", controls: "controls", autoplay: "autoplay", crossOrigin: "crossOrigin", height: "height", width: "width", loop: "loop", src: "src", volume: "volume", play: "play" }, viewQueries: [{ propertyName: "video", first: true, predicate: ["video"], descendants: true }, { propertyName: "progress", first: true, predicate: ["progress"], descendants: true }, { propertyName: "figure", first: true, predicate: ["figure"], descendants: true }], ngImport: i0, template: "<link href=\"https://fonts.googleapis.com/icon?family=Material+Icons\" rel=\"stylesheet\">\r\n<figure #figure [ngClass]=\"{'show-controls': mouseMoving || video.paused || video.ended}\">\r\n  <video #video id=\"video-tag\" preload=\"metadata\" [controls]=\"!browserSupportsVideo && _coerceBooleanProperty(controls)\"\r\n    [crossOrigin]=\"crossOrigin\" [autoplay]=\"autoplay\" [poster]=\"posterSrc\" [loop]=\"_coerceBooleanProperty(loop)\"\r\n    [volume]=\"volume\" (click)=\"toggleVideoPlayPause()\" (loadedmetadata)=\"onMetadataLodaded()\">\r\n    <source *ngFor=\"let source of _videoSources\" [src]=\"source.src\" [type]=\"source.type\">\r\n    <track *ngFor=\"let track of tracks\" [src]=\"track.src\" [srclang]=\"track.srclang\" [label]=\"track.label\"\r\n      [default]=\"track.default\">\r\n  </video>\r\n  <ul *ngIf=\"browserSupportsVideo && _coerceBooleanProperty(controls)\" class=\"controls\" aria-controls=\"video-tag\">\r\n    <li #progress tabindex=\"0\" class=\"progress\" (mousemove)=\"onProgressHover($event)\"\r\n      (keyup.ArrowRight)=\"advanceVideoBy(10)\" (keyup.ArrowLeft)=\"advanceVideoBy(-10)\">\r\n      <progress *ngFor=\"let chapter of chapters | sort; let i = index\" #chapterProgress tabindex=\"-1\"\r\n        [value]=\"video.currentTime - chapter.time\" [max]=\"chapter.duration || 0\"\r\n        [style.width.%]=\"(chapter.duration || 0) / video.duration * 100\" (click)=\"onProgressBarClick($event)\"\r\n        (keyup.ArrowRight)=\"advanceVideoBy(10)\" (keyup.ArrowLeft)=\"advanceVideoBy(-10)\" (focus)=\"onProgressHover()\">\r\n      </progress>\r\n      <div class=\"thumbnail\">\r\n        <img *ngIf=\"thumbnailSrc\" [src]=\"thumbnailSrc\" alt=\"\" crossorigin=\"anonymous\">\r\n        <p style=\"color: white\" *ngIf=\"thumbnailTime\">{{ thumbnailTime | chapter:chapters }}</p>\r\n        <p style=\"color: white\" *ngIf=\"thumbnailTime\">{{ thumbnailTime | time }}</p>\r\n      </div>\r\n      <div class=\"chapters\"></div>\r\n    </li>\r\n    <div class=\"buttons\">\r\n      <li><button type=\"button\" (click)=\"toggleVideoPlayPause()\">\r\n          <span *ngIf=\"video.paused || video.ended; else pauseIcon\" class=\"material-icons\">play_arrow</span>\r\n          <ng-template #pauseIcon><span class=\"material-icons\">pause</span></ng-template>\r\n        </button></li>\r\n      <li class=\"mute\"><button type=\"button\" (click)=\"video.muted = !video.muted\">\r\n          <span *ngIf=\"!video.muted && video.volume; else muteIcon\" class=\"material-icons\">volume_up</span>\r\n          <ng-template #muteIcon><span class=\"material-icons\">volume_off</span></ng-template>\r\n        </button></li>\r\n      <li class=\"volume-control\"><input type=\"range\" min=\"0\" max=\"1\" step=\"0.1\" (input)=\"onVolumeChange($event)\"\r\n          [value]=\"video.volume\"></li>\r\n      <li *ngIf=\"!isNaN(video.currentTime) && !isNaN(video.duration)\" class=\"text\"><span>{{video.currentTime | time}} /\r\n          {{video.duration | time}}</span></li>\r\n      <li *ngIf=\"chapters.length > 1 || chapters[0].title\" class=\"text\"><span>&bull;</span><span\r\n          style=\"margin-left: 4px\">{{video.currentTime | chapter:chapters}} </span></li>\r\n      <span style=\"flex-grow: 1;\"></span>\r\n      <li *ngIf=\"tracks.length\" class=\"tracks-control\" (clickOutside)=\"trackMenuOpen = false\">\r\n        <button type=\"button\" (click)=\"trackMenuOpen = !trackMenuOpen\"><span\r\n            class=\"material-icons\">subtitles</span></button>\r\n        <ul class=\"tracks-menu\" [ngClass]=\"{'open': trackMenuOpen}\">\r\n          <li *ngFor=\"let track of tracks\"><button (click)=\"selectSubtitles(track)\">{{ track.label}}</button></li>\r\n          <li><button (click)=\"selectSubtitles(null)\">Off</button></li>\r\n        </ul>\r\n      </li>\r\n      <li><button *ngIf=\"fullScreenEnabled\" type=\"button\" (click)=\"toggleFullScreen()\"><span\r\n            class=\"material-icons\">fullscreen</span></button></li>\r\n    </div>\r\n  </ul>\r\n</figure>", styles: [":host{display:block}figure{margin:0;padding:0;height:100%;width:100%;position:relative}video{display:block;width:100%;height:100%;background:#000;object-fit:contain}figure.show-controls .controls{transform:scaleY(1)}.controls{transform:scaleY(0);transform-origin:bottom;position:absolute;list-style:none;margin:0;box-sizing:border-box;bottom:0;left:0;width:100%;background:linear-gradient(to top,#000,#0000001a);padding:16px 8px 4px;transition:transform .2s ease-in-out}.controls .buttons{display:flex}.controls ::ng-deep .material-icons{color:#fff;font-size:20px}.controls button{border:none;background:none;cursor:pointer;padding:0;margin:0;width:36px;height:36px;border-radius:50%;display:flex;align-items:center;justify-content:center}.controls button:focus,.controls button:hover{background:#ffffff1a}.progress{margin-bottom:4px;cursor:pointer;position:relative;display:flex}.progress:hover:before{content:var(--thumbnail-src);position:absolute;top:-100px;left:calc(var(--hover-x) - 80px);width:160px;height:min-content}.progress:hover:after{content:\"\";display:block;position:absolute;left:0;top:0;width:var(--hover-x);height:100%;background-color:#0000004d;overflow:hidden}.progress:hover progress{height:8px}.progress progress{flex-grow:1;flex-shrink:1;-webkit-appearance:none;appearance:none;height:4px;transition:height .2s ease-in-out;position:relative}.progress progress:not(:first-child){margin-left:4px}.progress:hover .thumbnail{visibility:visible}.thumbnail{display:block;visibility:hidden;position:absolute;width:160px;left:calc(var(--hover-x) - 80px);transform:translateY(-100%);padding-bottom:8px}.thumbnail img{border:2px solid white;display:block}.thumbnail p{text-align:center;margin-top:0;margin-bottom:0}.text{display:flex;justify-content:center;align-items:center;color:#fff;margin-left:8px}.volume-control{display:flex;justify-content:center;align-items:center;width:0;overflow:hidden;transition:width .2s ease-in-out}.volume-control input{cursor:pointer;-webkit-appearance:none;appearance:none;background:#fff;height:4px;border-radius:2px;width:80px}.volume-control input::-webkit-slider-thumb{-webkit-appearance:none;appearance:none;width:12px;height:12px;border-radius:6px;background:#fff}.mute:hover+.volume-control,.volume-control:hover{width:80px}.tracks-control{position:relative}.tracks-menu{position:absolute;top:0;left:0;transform:translateY(calc(-100% - 16px));list-style-type:none;background-color:#000;padding:0;display:none;opacity:0;transition:opacity .2s ease-in-out}.tracks-menu button{width:100%;border-radius:0;color:#fff;display:block;padding-inline:8px;font-weight:700}.tracks-menu.open{opacity:1;display:block}\n"], dependencies: [{ kind: "directive", type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: ClickOutsideDirective, selector: "[clickOutside]", outputs: ["clickOutside"] }, { kind: "pipe", type: TimePipe, name: "time" }, { kind: "pipe", type: ChapterPipe, name: "chapter" }, { kind: "pipe", type: SortPipe, name: "sort" }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: NgxVideoPlayerComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ngx-video-player', standalone: false, template: "<link href=\"https://fonts.googleapis.com/icon?family=Material+Icons\" rel=\"stylesheet\">\r\n<figure #figure [ngClass]=\"{'show-controls': mouseMoving || video.paused || video.ended}\">\r\n  <video #video id=\"video-tag\" preload=\"metadata\" [controls]=\"!browserSupportsVideo && _coerceBooleanProperty(controls)\"\r\n    [crossOrigin]=\"crossOrigin\" [autoplay]=\"autoplay\" [poster]=\"posterSrc\" [loop]=\"_coerceBooleanProperty(loop)\"\r\n    [volume]=\"volume\" (click)=\"toggleVideoPlayPause()\" (loadedmetadata)=\"onMetadataLodaded()\">\r\n    <source *ngFor=\"let source of _videoSources\" [src]=\"source.src\" [type]=\"source.type\">\r\n    <track *ngFor=\"let track of tracks\" [src]=\"track.src\" [srclang]=\"track.srclang\" [label]=\"track.label\"\r\n      [default]=\"track.default\">\r\n  </video>\r\n  <ul *ngIf=\"browserSupportsVideo && _coerceBooleanProperty(controls)\" class=\"controls\" aria-controls=\"video-tag\">\r\n    <li #progress tabindex=\"0\" class=\"progress\" (mousemove)=\"onProgressHover($event)\"\r\n      (keyup.ArrowRight)=\"advanceVideoBy(10)\" (keyup.ArrowLeft)=\"advanceVideoBy(-10)\">\r\n      <progress *ngFor=\"let chapter of chapters | sort; let i = index\" #chapterProgress tabindex=\"-1\"\r\n        [value]=\"video.currentTime - chapter.time\" [max]=\"chapter.duration || 0\"\r\n        [style.width.%]=\"(chapter.duration || 0) / video.duration * 100\" (click)=\"onProgressBarClick($event)\"\r\n        (keyup.ArrowRight)=\"advanceVideoBy(10)\" (keyup.ArrowLeft)=\"advanceVideoBy(-10)\" (focus)=\"onProgressHover()\">\r\n      </progress>\r\n      <div class=\"thumbnail\">\r\n        <img *ngIf=\"thumbnailSrc\" [src]=\"thumbnailSrc\" alt=\"\" crossorigin=\"anonymous\">\r\n        <p style=\"color: white\" *ngIf=\"thumbnailTime\">{{ thumbnailTime | chapter:chapters }}</p>\r\n        <p style=\"color: white\" *ngIf=\"thumbnailTime\">{{ thumbnailTime | time }}</p>\r\n      </div>\r\n      <div class=\"chapters\"></div>\r\n    </li>\r\n    <div class=\"buttons\">\r\n      <li><button type=\"button\" (click)=\"toggleVideoPlayPause()\">\r\n          <span *ngIf=\"video.paused || video.ended; else pauseIcon\" class=\"material-icons\">play_arrow</span>\r\n          <ng-template #pauseIcon><span class=\"material-icons\">pause</span></ng-template>\r\n        </button></li>\r\n      <li class=\"mute\"><button type=\"button\" (click)=\"video.muted = !video.muted\">\r\n          <span *ngIf=\"!video.muted && video.volume; else muteIcon\" class=\"material-icons\">volume_up</span>\r\n          <ng-template #muteIcon><span class=\"material-icons\">volume_off</span></ng-template>\r\n        </button></li>\r\n      <li class=\"volume-control\"><input type=\"range\" min=\"0\" max=\"1\" step=\"0.1\" (input)=\"onVolumeChange($event)\"\r\n          [value]=\"video.volume\"></li>\r\n      <li *ngIf=\"!isNaN(video.currentTime) && !isNaN(video.duration)\" class=\"text\"><span>{{video.currentTime | time}} /\r\n          {{video.duration | time}}</span></li>\r\n      <li *ngIf=\"chapters.length > 1 || chapters[0].title\" class=\"text\"><span>&bull;</span><span\r\n          style=\"margin-left: 4px\">{{video.currentTime | chapter:chapters}} </span></li>\r\n      <span style=\"flex-grow: 1;\"></span>\r\n      <li *ngIf=\"tracks.length\" class=\"tracks-control\" (clickOutside)=\"trackMenuOpen = false\">\r\n        <button type=\"button\" (click)=\"trackMenuOpen = !trackMenuOpen\"><span\r\n            class=\"material-icons\">subtitles</span></button>\r\n        <ul class=\"tracks-menu\" [ngClass]=\"{'open': trackMenuOpen}\">\r\n          <li *ngFor=\"let track of tracks\"><button (click)=\"selectSubtitles(track)\">{{ track.label}}</button></li>\r\n          <li><button (click)=\"selectSubtitles(null)\">Off</button></li>\r\n        </ul>\r\n      </li>\r\n      <li><button *ngIf=\"fullScreenEnabled\" type=\"button\" (click)=\"toggleFullScreen()\"><span\r\n            class=\"material-icons\">fullscreen</span></button></li>\r\n    </div>\r\n  </ul>\r\n</figure>", styles: [":host{display:block}figure{margin:0;padding:0;height:100%;width:100%;position:relative}video{display:block;width:100%;height:100%;background:#000;object-fit:contain}figure.show-controls .controls{transform:scaleY(1)}.controls{transform:scaleY(0);transform-origin:bottom;position:absolute;list-style:none;margin:0;box-sizing:border-box;bottom:0;left:0;width:100%;background:linear-gradient(to top,#000,#0000001a);padding:16px 8px 4px;transition:transform .2s ease-in-out}.controls .buttons{display:flex}.controls ::ng-deep .material-icons{color:#fff;font-size:20px}.controls button{border:none;background:none;cursor:pointer;padding:0;margin:0;width:36px;height:36px;border-radius:50%;display:flex;align-items:center;justify-content:center}.controls button:focus,.controls button:hover{background:#ffffff1a}.progress{margin-bottom:4px;cursor:pointer;position:relative;display:flex}.progress:hover:before{content:var(--thumbnail-src);position:absolute;top:-100px;left:calc(var(--hover-x) - 80px);width:160px;height:min-content}.progress:hover:after{content:\"\";display:block;position:absolute;left:0;top:0;width:var(--hover-x);height:100%;background-color:#0000004d;overflow:hidden}.progress:hover progress{height:8px}.progress progress{flex-grow:1;flex-shrink:1;-webkit-appearance:none;appearance:none;height:4px;transition:height .2s ease-in-out;position:relative}.progress progress:not(:first-child){margin-left:4px}.progress:hover .thumbnail{visibility:visible}.thumbnail{display:block;visibility:hidden;position:absolute;width:160px;left:calc(var(--hover-x) - 80px);transform:translateY(-100%);padding-bottom:8px}.thumbnail img{border:2px solid white;display:block}.thumbnail p{text-align:center;margin-top:0;margin-bottom:0}.text{display:flex;justify-content:center;align-items:center;color:#fff;margin-left:8px}.volume-control{display:flex;justify-content:center;align-items:center;width:0;overflow:hidden;transition:width .2s ease-in-out}.volume-control input{cursor:pointer;-webkit-appearance:none;appearance:none;background:#fff;height:4px;border-radius:2px;width:80px}.volume-control input::-webkit-slider-thumb{-webkit-appearance:none;appearance:none;width:12px;height:12px;border-radius:6px;background:#fff}.mute:hover+.volume-control,.volume-control:hover{width:80px}.tracks-control{position:relative}.tracks-menu{position:absolute;top:0;left:0;transform:translateY(calc(-100% - 16px));list-style-type:none;background-color:#000;padding:0;display:none;opacity:0;transition:opacity .2s ease-in-out}.tracks-menu button{width:100%;border-radius:0;color:#fff;display:block;padding-inline:8px;font-weight:700}.tracks-menu.open{opacity:1;display:block}\n"] }]
        }], propDecorators: { posterSrc: [{
                type: Input
            }], chapters: [{
                type: Input
            }], tracks: [{
                type: Input
            }], sources: [{
                type: Input
            }], controls: [{
                type: Input
            }], autoplay: [{
                type: Input
            }], crossOrigin: [{
                type: Input
            }], height: [{
                type: Input
            }], width: [{
                type: Input
            }], loop: [{
                type: Input
            }], src: [{
                type: Input
            }], volume: [{
                type: Input
            }], play: [{
                type: Input
            }], video: [{
                type: ViewChild,
                args: ['video']
            }], progress: [{
                type: ViewChild,
                args: ['progress']
            }], figure: [{
                type: ViewChild,
                args: ['figure']
            }] } });

class NgxVideoPlayerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: NgxVideoPlayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "19.2.5", ngImport: i0, type: NgxVideoPlayerModule, declarations: [NgxVideoPlayerComponent,
            ClickOutsideDirective,
            TimePipe,
            ChapterPipe,
            SortPipe], imports: [CommonModule], exports: [NgxVideoPlayerComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: NgxVideoPlayerModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "19.2.5", ngImport: i0, type: NgxVideoPlayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxVideoPlayerComponent,
                        ClickOutsideDirective,
                        TimePipe,
                        ChapterPipe,
                        SortPipe,
                    ],
                    imports: [CommonModule],
                    exports: [NgxVideoPlayerComponent],
                }]
        }] });

/*
 * Public API Surface of ngx-video-player
 */

/**
 * Generated bundle index. Do not edit.
 */

export { NgxVideoPlayerComponent, NgxVideoPlayerModule };
