// angular-video-js-player/fesm2015/angular-video-js-player.mjs
import { NgFor } from '@angular/common';
import * as i0 from '@angular/core';
import { Component, Input } from '@angular/core';

var PlayerKey;
(function (PlayerKey) {
    PlayerKey["arrowLeft"] = "ArrowLeft";
    PlayerKey["arrowRight"] = "ArrowRight";
    PlayerKey["arrowDown"] = "ArrowDown";
    PlayerKey["arrowUp"] = "ArrowUp";
    PlayerKey["space"] = "Space";
})(PlayerKey || (PlayerKey = {}));
class VideoJsComponent {
    constructor() {
        this.videoUrl = [];
        this.captionList = [];
        this.qualitySelectorUrl = 'https://unpkg.com/@silvermine/videojs-quality-selector/dist/js/silvermine-videojs-quality-selector.min.js';
        this.videoJsUrl = 'https://vjs.zencdn.net/7.12.3/video.min.js';
        this.loadScript();
    }
    createScript(url, cb = () => { }) {
        const body = document.body;
        const script = document.createElement('script');
        script.innerHTML = '';
        script.addEventListener("load", () => cb());
        script.src = url;
        script.defer = true;
        body.appendChild(script);
    }
    loadScript() {
        const vjs = this.videoJsUrl;
        this.createScript(vjs, () => {
            this.createScript(this.qualitySelectorUrl);
            this.setTimer = setTimeout(() => {
                this.onPlayVideo();
            }, 300);
        });
    }
    onPlayVideo() {
        var _a;
        const options = this.setVideoConfig();
        this.player = videojs('video-player', options);
        this.qualitySelector && !((_a = this.player.controlBar.childNameIndex_) === null || _a === void 0 ? void 0 : _a.QualitySelector) && this.player.controlBar.addChild('QualitySelector');
    }
    setVideoConfig() {
        return {
            preload: "auto",
            autoplay: true,
            controls: true,
            responsive: true,
            html5: {
                nativeTextTracks: false,
            },
            userActions: {
                hotkeys: (event) => {
                    this.onPressHotkeys(event);
                },
                doubleClick: () => {
                    if (this.player.isFullscreen() && this.fullScreenModeEnable) {
                        this.player.exitFullscreen();
                    }
                    else {
                        this.fullScreenModeEnable && this.player.requestFullscreen();
                    }
                },
            },
        };
    }
    onPressHotkeys(event) {
        if (event.code === PlayerKey.space && this.playPauseByKey) {
            if (this.player.paused()) {
                this.player.play();
            }
            else {
                this.player.pause();
            }
        }
        if (event.code === PlayerKey.arrowLeft && this.volumeEnableByKey) {
            const backward = this.player.currentTime() - 10;
            this.player.currentTime(backward);
        }
        if (event.code === PlayerKey.arrowRight && this.volumeEnableByKey) {
            const forward = this.player.currentTime() + 10;
            this.player.currentTime(forward);
        }
        if (event.code === PlayerKey.arrowUp && this.volumeEnableByKey) {
            this.player.volume(this.player.volume() + 0.1);
        }
        if (event.code === PlayerKey.arrowDown && this.volumeEnableByKey) {
            this.player.volume(this.player.volume() - 0.1);
        }
    }
    ngOnDestroy() {
        this.player.dispose();
        clearTimeout(this.setTimer);
    }
}
VideoJsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: VideoJsComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
VideoJsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: VideoJsComponent, isStandalone: true, selector: "lib-video-js", inputs: { videoUrl: "videoUrl", captionList: "captionList", volumeEnableByKey: "volumeEnableByKey", playPauseByKey: "playPauseByKey", fullScreenModeEnable: "fullScreenModeEnable", qualitySelector: "qualitySelector" }, ngImport: i0, template: `
 <video id="video-player" class="video-js vjs-default-skin" autoplay="true">
  <source *ngFor="let item of videoUrl" src="{{ item.url }}" type="application/x-mpegURL" [attr.label]="item.label??''" />
  <ng-container *ngFor="let caption of captionList">
    <track kind="captions" src="{{caption.url}}" srclang="{{caption?.lang}}" label="{{caption?.label}}" />
  </ng-container>
 </video>
  `, isInline: true, dependencies: [{ kind: "directive", type: NgFor, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: VideoJsComponent, decorators: [{
            type: Component,
            args: [{ selector: 'lib-video-js', standalone: true, imports: [NgFor], template: `
 <video id="video-player" class="video-js vjs-default-skin" autoplay="true">
  <source *ngFor="let item of videoUrl" src="{{ item.url }}" type="application/x-mpegURL" [attr.label]="item.label??''" />
  <ng-container *ngFor="let caption of captionList">
    <track kind="captions" src="{{caption.url}}" srclang="{{caption?.lang}}" label="{{caption?.label}}" />
  </ng-container>
 </video>
  ` }]
        }], ctorParameters: function () { return []; }, propDecorators: { videoUrl: [{
                type: Input
            }], captionList: [{
                type: Input
            }], volumeEnableByKey: [{
                type: Input
            }], playPauseByKey: [{
                type: Input
            }], fullScreenModeEnable: [{
                type: Input
            }], qualitySelector: [{
                type: Input
            }] } });

/*
 * Public API Surface of video-js
 */

/**
 * Generated bundle index. Do not edit.
 */

export { VideoJsComponent };
